nixos/stage1: copy initrd secrets into place after special mounts

ddz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This modifies `initialRamdiskSecretAppender` to stage secrets in `/initrd-secrets/` and `stage-1-init` to copy them into place after mounting special file systems. This allows secrets to be copied into `ramfs` mounts like `/run/keys` for use after `stage-1` finishes without copying them to disk (which would not be very secure). This is useful for ZFS filesystem key files, which are mounted later, as well as when rebuilding system configurations to be able to pull secret from `/run/keys` to rebuild the initrd.

###### Things done

Tested on `release-21.05` branch and `master` to preserve existing functionality as well as enable a destinations for secrets in `/run/keys` which were present after system had booted. 

Updated test in `tests/initrd-secrets.nix` to verify new behavior.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ddz committed:
nixos/stage1: copy initrd secrets into place after special mounts

This modifies initialRamdiskSecretAppender to stage secrets in
/.initrd-secrets/ and stage-1-init to copy them into place after mounting
special file systems. This allows secrets to be copied into ramfs mounts
like /run/keys for use after stage-1 finishes without copying them to disk
(which would not be very secure).

SuperSandro2000 commented on nixos/modules/system/boot/stage-1-init.sh:
```diff
@@ -119,6 +119,14 @@ specialMount() {
+if [ -d "/initrd-secrets" ]; then
+    for secret in $(cd "/initrd-secrets"; find . -type f); do
+        mkdir -p $(dirname "/$secret")
+	cp "/initrd-secrets/$secret" "$secret"
```
```suggestion
    cp "/initrd-secrets/$secret" "$secret"
```

ddz replied:
This has been addressed, resolving conversation.

SuperSandro2000 commented on nixos/modules/system/boot/stage-1-init.sh:
```diff
@@ -119,6 +119,14 @@ specialMount() {
 
+# Copy initrd secrets staged into /initrd-secrets to their real destinations
+if [ -d "/initrd-secrets" ]; then
+    for secret in $(cd "/initrd-secrets"; find . -type f); do
```
```suggestion
    for secret in $(find /initrd-secrets -type f); do
```

ddz replied:
While awkward, this form is needed because [boot.initrd.secrets](https://nixos.org/manual/nixos/stable/options.html#opt-boot.initrd.secrets) takes full paths for the attribute name and this change roots them under `/initrd-secrets` (e.g. `/initrd-secrets/etc/dropbear/dropbear_rsa_host_key`).

ddz replied:
This has been addressed with more clarifying comments, resolving conversation...

ddz commented:
Thanks for the review! I incorporated your suggestions and squashed back into this PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/559

ddz committed:
nixos/tests: update initrd-secrets test to test secret in /run/keys

Since /run/keys is a ramfs, it is not paged out and a good place to copy
secrets to. Test whether secrets with a path in /run/keys exist after initrd.

ddz commented:
cc: @Mic92 - you reviewed one of my PRs a few years back, would you mind taking a look at this small one?

dasJ commented:
How would that approach be secure in any way? Your initrd is usually unencrypted and so are the secrets inside. Since your secrets are already in your unencrypted initrd in your unencrypted /boot, why not store them in your regular filesystem as well?

ddz commented:
> How would that approach be secure in any way? Your initrd is usually unencrypted and so are the secrets inside. Since your secrets are already in your unencrypted initrd in your unencrypted /boot, why not store them in your regular filesystem as well?

Thank you for the review! This PR doesn't change how secrets are stored or where only modifying them in a way to permit the system configuration to keep a copy on one of the special file systems after the initrd is done (to have, for example a clean contract between initrd and runtime where needed secrets are available in `/run`). In the case of my systems, I have a LUKS-encrypted /boot, which keeps the secrets initrd encrypted when powered off and I'm (slowly, sporadically) working on something which keeps them encrypted (e.g. TPM sealed) on the /boot filesystem as well.

dasJ commented on nixos/modules/system/boot/stage-1-init.sh:
```diff
@@ -119,6 +119,18 @@ specialMount() {
+    # Secrets are named by their full destination pathname and stored
+    # under /.initrd-secrets/
+    #
+    for secret in $(cd "/.initrd-secrets"; find . -type f); do
```
```suggestion
    for secret in $(find /.initrd-secrets -not -type d); do
```

ddz replied:
Hello! @SuperSandro2000 [suggested the same thing here](https://github.com/NixOS/nixpkgs/pull/129449#discussion_r664689775), but that changes the output and paths assigned to `secret` (they would have `/.initrd-secrets/` still prepended vs. being the full path specified as the attribute name to `boot.initrd.secrets`. I chose this approach rather than having to strip the first path component inside the loop, which would be less precise. In addition, `-not -type d` is not equivalent to explicitly including files (`-type f`), it could include links, sockets, device nodes, etc. 

I tried to explain the above in the comment preceding the loop, but if it's still not clear to reviewers, I should try to clarify more.

dasJ commented:
I still don't see any point in having "secrets" that are in the plain initrd but there's really also no reason we shouldn't allow for this if this is functionality people want to have

ddz commented:
Thank you for the merge!
